### ISSUE TITLE
Inconsistent return type for det(::Hermitian) and logdet(::Hermitian)

### ISSUE THREAD

**Issue by filtron** [JuliaLang/julia#46332]
Since Hermitian matrices have strictly real eigenvalues one would expect both these functions to have a return type of <: Real. 
In practice this is only the case for det but not for logdet. 

minimal working example: 

```
julia> L = randn(ComplexF64,2,2) 
2×2 Matrix{ComplexF64}:
 0.248839+0.558406im  -1.35699+0.329374im
 0.888645-0.6176im    -1.31893-1.2155im

julia> P = Hermitian(L*L')
2×2 Hermitian{ComplexF64, Matrix{ComplexF64}}:
 2.32364+0.0im      1.26568-1.43393im
 1.26568+1.43393im  4.38815+0.0im

julia> det(P)::Real
6.538363080754605

julia> logdet(P)::Real
ERROR: TypeError: in typeassert, expected Real, got a value of type ComplexF64
Stacktrace:
 [1] top-level scope
   @ REPL[4]:1

julia> versioninfo()
Julia Version 1.7.1
Commit ac5cc99908 (2021-12-22 19:35 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6800K CPU @ 3.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, broadwell)
```

**Comment by jishnub** [comment]
`log` of real numbers aren't necessarily real, though. Negative determinants would lead to complex `logdet`.
```julia
julia> A = Hermitian(Matrix(Diagonal(ComplexF64[1, -0.5])))
2×2 Hermitian{ComplexF64, Matrix{ComplexF64}}:
 1.0+0.0im   0.0+0.0im
 0.0-0.0im  -0.5+0.0im

julia> det(A)
-0.5

julia> logdet(A)
-0.6931471805599453 + 3.141592653589793im
```

**Comment by mikmoore** [comment]
You may be interested in `logabsdet`.

Orthogonal to that, there does appear to be something very wrong on `v1.8.0-rc1`. Can somebody check this on a more recent version?
```julia
julia> versioninfo()
Julia Version 1.8.0-rc1
Commit 6368fdc656 (2022-05-27 18:33 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: 12 × Intel(R) Core(TM) i7-9850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 12 virtual cores

julia> @show P = Hermitian(randn(ComplexF64,3,3));
P = Hermitian(randn(ComplexF64, 3, 3)) = ComplexF64[0.5175918433201382 + 0.0im 0.07699799944173633 - 0.1777444666002341im 0.9230890447699571 - 0.6154956708992522im; 0.07699799944173633 + 0.1777444666002341im -0.8287684406666638 + 0.0im 0.5478700511035655 - 0.8197950686892131im; 0.9230890447699571 + 0.6154956708992522im 0.5478700511035655 + 0.8197950686892131im -0.678521284911168 + 0.0im]

julia> det(P)
0.8399040378129923

julia> log(det(P))
-0.17446763436956214

julia> logdet(P)
0.12793427559134427 + 0.8218324882522688im

julia> logabsdet(P)
(0.1279342755913443, 0.6808802464402095 + 0.7323947637766941im)
```
`logdet` and `logabsdet` both give totally wrong results.

**Comment by filtron** [comment]
> `log` of real numbers aren't necessarily real, though. Negative determinants would lead to complex `logdet`.
> 
> ```julia
> julia> A = Hermitian(Matrix(Diagonal(ComplexF64[1, -0.5])))
> 2×2 Hermitian{ComplexF64, Matrix{ComplexF64}}:
>  1.0+0.0im   0.0+0.0im
>  0.0-0.0im  -0.5+0.0im
> 
> julia> det(A)
> -0.5
> 
> julia> logdet(A)
> -0.6931471805599453 + 3.141592653589793im
> ```

hm... unfortunately I do not think it is this simple. According to the documentation ```logdet(M)``` and ```log(det(M))``` are supposed to be equivalent (https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.logdet). But in your example they are clearly not, i.e:

```julia 
julia> A = Hermitian(Matrix(Diagonal(ComplexF64[1, -0.5])))
2×2 Hermitian{ComplexF64, Matrix{ComplexF64}}:
 1.0+0.0im   0.0+0.0im
 0.0-0.0im  -0.5+0.0im

julia> logdet(A)
-0.6931471805599453 + 3.141592653589793im

julia> log(det(A))
ERROR: DomainError with -0.5:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(f::Symbol, x::Float64)
   @ Base.Math ./math.jl:33
 [2] _log(x::Float64, base::Val{:ℯ}, func::Symbol)
   @ Base.Math ./special/log.jl:304
 [3] log(x::Float64)
   @ Base.Math ./special/log.jl:269
 [4] top-level scope
   @ REPL[3]:1
```

**Comment by filtron** [comment]
> `log` of real numbers aren't necessarily real, though. Negative determinants would lead to complex `logdet`.
> 
> ```julia
> julia> A = Hermitian(Matrix(Diagonal(ComplexF64[1, -0.5])))
> 2×2 Hermitian{ComplexF64, Matrix{ComplexF64}}:
>  1.0+0.0im   0.0+0.0im
>  0.0-0.0im  -0.5+0.0im
> 
> julia> det(A)
> -0.5
> 
> julia> logdet(A)
> -0.6931471805599453 + 3.141592653589793im
> ```

hm... unfortunately I do not think it is this simple. According to the documentation ```logdet(M)``` and ```log(det(M))``` are supposed to be equivalent (https://docs.julialang.org/en/v1/stdlib/LinearAlgebra/#LinearAlgebra.logdet). But in your example they are clearly not, i.e:

```julia 
julia> A = Hermitian(Matrix(Diagonal(ComplexF64[1, -0.5])))
2×2 Hermitian{ComplexF64, Matrix{ComplexF64}}:
 1.0+0.0im   0.0+0.0im
 0.0-0.0im  -0.5+0.0im

julia> logdet(A)
-0.6931471805599453 + 3.141592653589793im

julia> log(det(A))
ERROR: DomainError with -0.5:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
Stacktrace:
 [1] throw_complex_domainerror(f::Symbol, x::Float64)
   @ Base.Math ./math.jl:33
 [2] _log(x::Float64, base::Val{:ℯ}, func::Symbol)
   @ Base.Math ./special/log.jl:304
 [3] log(x::Float64)
   @ Base.Math ./special/log.jl:269
 [4] top-level scope
   @ REPL[3]:1
```

**Comment by jishnub** [comment]
> ```julia
> julia> @show P = Hermitian(randn(ComplexF64,3,3));
> P = Hermitian(randn(ComplexF64, 3, 3)) = ComplexF64[0.5175918433201382 + 0.0im 0.07699799944173633 - 0.1777444666002341im 0.9230890447699571 - 0.6154956708992522im; 0.07699799944173633 + 0.1777444666002341im -0.8287684406666638 + 0.0im 0.5478700511035655 - 0.8197950686892131im; 0.9230890447699571 + 0.6154956708992522im 0.5478700511035655 + 0.8197950686892131im -0.678521284911168 + 0.0im]
> 
> julia> det(P)
> 0.8399040378129923
> 
> julia> log(det(P))
> -0.17446763436956214
> 
> julia> logdet(Matrix(P))
> -0.17446763436956197 - 1.6653345369377346e-16im
> 
> julia> logdet(P)
> 0.12793427559134427 + 0.8218324882522688im
> 
> julia> logabsdet(P)
> (0.1279342755913443, 0.6808802464402095 + 0.7323947637766941im)
> ```

This seems fixed on `v1.8.0-rc4`. I've copied the matrix from your output
```julia
julia> P = ComplexF64[0.5175918433201382 + 0.0im 0.07699799944173633 - 0.1777444666002341im 0.9230890447699571 - 0.6154956708992522im; 0.07699799944173633 + 0.1777444666002341im -0.8287684406666638 + 0.0im 0.5478700511035655 - 0.8197950686892131im; 0.9230890447699571 + 0.6154956708992522im 0.5478700511035655 + 0.8197950686892131im -0.678521284911168 + 0.0im]
3×3 Matrix{ComplexF64}:
 0.517592+0.0im        0.076998-0.177744im   0.923089-0.615496im
 0.076998+0.177744im  -0.828768+0.0im         0.54787-0.819795im
 0.923089+0.615496im    0.54787+0.819795im  -0.678521+0.0im

julia> A = Hermitian(P);

julia> log(det(A))
-0.17446763436956214

julia> logdet(A)
-0.17446763436956197 - 1.6653345369377346e-16im

julia> logabsdet(A)
(-0.1744676343695622, 1.0000000000000002 - 1.6653345369377348e-16im)

julia> versioninfo()
Julia Version 1.8.0-rc4
Commit 7853436ccd7 (2022-08-08 07:55 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 8 × Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 8 virtual cores
Environment:
  JULIA_EDITOR = vim
```

**Comment by dkarrasch** [comment]
IIUC, then https://github.com/JuliaLang/julia/issues/46332#issuecomment-1213380909 explains the behaviour, and the other issue was fixed, so I'm closing this. Feel free to reopen if something was unadressed.
